### PR TITLE
feat(compiler): access modifiers

### PR DESCRIPTION
**Pull request by yoav-steinberg**
See #108 
This PR adds `public` and `protected` support to class members and while doing so makes all other members **private** by default. It also verifies interface method implementations must be `public`. It also includes support for JSII access modifiers. To get this working I did the following refactors:

* Cleaned up fields in the `SymbolEnv` and instead added a `kind` field distinguishing between 3 types of environments: a simple code block (`Scope`) a function/method/closure environemt (`Function`) and an environment used to store members of a type (`Type`). This was primarily needed so I can use the result of a symbol env lookup to figure out in what ancestor class a given member was defined. The added benefit was code clarity and sanity checks which, btw, found a non-related bug in super constructor type checking (see below). One piece of such clarity was the long awaited removal of `return_type` from `SymbolEnv`.
* Cleaned up the `ContextVisitor` which I needed to add to the `TypeChecker`. Here I removed the requirement to store the `init_env` in `push_class` which didn't make sense, and is probably a leftover of some iteration on our lifting code. I also cleaned up the `push_method` code to make more sense and now it's a more general `push_function` stack with a helper for when we're just interested in methods.
* I added some more stuff to `VisitorWithContext` trait, for easy "push -> do stuff or even return -> pop" pattern. Specifically I needed the `TypeChecker` to push function definitions and current statement safely.
* Use a `VisitorContext` in `TypeChecker` to track context so I know in what class I am for access modifier checking. This approach also provides a better way of tracking the current `return_type` so I could remove it from the env. This also removed all sorts of dirt from the `TypeChecker` struct (like `in_json`, etc.).
* The `type_check_statement` method was getting way to big, so I moved the contents of its match statements into separate methods.
* The sanity checks when creating new `SymbolEnv`s detected that we're creating a weird env in `type_check_super_constructor_against_parent_initializer`. I removed a large bunch of code there that didn't seem correct to me.
* The actual access modifier verification takes place inside `get_property_from_class_like` (**<= this is where you want to look**). This way I avoid cascading errors and "unhydrated" classes. To make this work for all cases I updated `resolve_referece` for static type references to also use this method instead of doing its own thing. I think this is cleaner.

Still left and possible candidate for followup PRs:
- [x] Tests for JSII imports with access modifiers
- [x] Handle method overriding rules
- [x] Support `public` for types (export types from module) -> out of scope, will be handled added by our module system.
- [x] Decide if we allow `public` fields: tracking issue #4180
- [x] Decide about `internal`: tracking issue #4156

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingc/src/type_check.rs [hunk 1]
```diff
@@ -3770,24 +3924,22 @@ impl<'a> TypeChecker<'a> {
...
+			SymbolEnvKind::Scope,
 			env.phase,
 			stmt.idx,
 		));
 		self.types.set_scope_env(statements, if_scope_env);
-		self.inner_scopes.push(statements);
+		self.inner_scopes.push((statements, self.ctx.clone()));
 	}
 
 	fn type_check_super_constructor_against_parent_initializer(
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:3935]: @hasanaburayyan please take a look at my changes here, just deleted some code that didn't make sense to me. Am I missing something?

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 2]
```diff
@@ -10,6 +10,105 @@ exports[`access_hidden_namespace.w 1`] = `
...
+Test Files 1 failed (1)
+Duration <DURATION>"
+`;
+
+exports[`access_modifiers.w 1`] = `
+"error: Cannot access protected member protected_field of Foo
+   --> ../../../examples/tests/invalid/access_modifiers.w:64:9
+   |
+64 | log(foo.protected_field);
+   |         ^^^^^^^^^^^^^^^ Cannot access protected member protected_field of Foo
```
Comment (Chriscbr) [inline tools/hangar/__snapshots__/invalid.ts.snap:23]: Should we put names of user identifiers in quotes or backticks to signal they're not part of the error message? e.g.
```suggestion
   |         ^^^^^^^^^^^^^^^ Cannot access protected member "protected_field" of "Foo"
```

I don't think anyone has proposed a style guide yet (https://github.com/winglang/wing/issues/1746) but I think the compiler already does this for some other error messages

File: libs/wingc/src/type_check.rs [hunk 3]
```diff
@@ -2944,764 +2967,890 @@ impl<'a> TypeChecker<'a> {
...
-				// preflight classes cannot be declared inside an inflight scope
-				// (the other way is okay)
-				if env.phase == Phase::Inflight && *phase == Phase::Preflight {
-					self.spanned_error(stmt, format!("Cannot declare a {} class in {} scope", phase, env.phase));
-				}
-				// Verify parent is a known class and get their env
-				let (parent_class, parent_class_env) = self.extract_parent_class(parent.as_ref(), *phase, name, env);
+			StmtKind::Class(ast_class) => {
+				tc.type_check_class(env, stmt, ast_class);
+			}
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:3033]: Great cleanup

File: libs/wingc/src/type_check.rs [hunk 4]
```diff
@@ -4728,20 +4852,60 @@ impl<'a> TypeChecker<'a> {
...
-		if let LookupResult::Found(field, _) = lookup_res {
+		if let LookupResult::Found(field, lookup_info) = lookup_res {
 			let var = field.as_variable().expect("Expected property to be a variable");
+
+			// Determine the access type of the property
+			// Lookup the property in the class env to find out in which class (perhaps an ancestor) it was defined
+			let SymbolEnvKind::Type(property_defined_in) = lookup_info.env.kind else {
+				panic!("Expected env to be a type env");
+			};
+			// Check if the class in which the property is defined is the current class nesting
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:4872]: What does this sentence mean?

File: examples/tests/invalid/access_modifiers.w [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  public public_method() {}
+  protected protected_method() {}
+  private_method() {}
+
+  public static public_static_method() {}
+  protected static protected_static_method() {}
+  static private_static_method() {}
+}
+
+class Bar extends Foo {
```
Comment (Chriscbr) [inline examples/tests/invalid/access_modifiers.w:40]: Can we add a test checking this works with more than one level of inheritance?

File: examples/tests/invalid/access_modifiers.w [hunk 6]
```diff
@@ -0,0 +1,100 @@
+class Foo {
+  protected protected_field: str;
+  private_field: str;
+  public public_field: str;
```
Comment (Chriscbr) [inline examples/tests/invalid/access_modifiers.w:4]: Can access modifiers be added to an interface? (add a valid / invalid unit test)

**Comment by Chriscbr** [comment]
Looks really great so far!

File: libs/wingc/src/type_check.rs (hunk 1 above)
Comment (hasanaburayyan) [reply libs/wingc/src/type_check.rs:3927]: Will do, excited to review this 😁

File: libs/wingc/src/type_check.rs [hunk 7]
```diff
@@ -4520,6 +4638,13 @@ impl<'a> TypeChecker<'a> {
 
 				let mut property_variable = self.resolve_variable_from_instance_type(instance_type, property, env, object);
 
+				// Validate access modifier
+				// if !property_variable.type_.is_unresolved() {
+				// 	// TODO: this isn't good enough becaues `instance_type` might be an array so we need to "actual" class_type (Array). This is done for us in
```
Comment (hasanaburayyan) [inline libs/wingc/src/type_check.rs:4643]: do we have a tracking issue?

**Review by hasanaburayyan — COMMENTED** [review]
Love the cleanups, let a comment about a bug with super initializers.

File: libs/tree-sitter-wing/grammar.js [hunk 8]
```diff
@@ -532,7 +532,7 @@ module.exports = grammar({
 
     async_modifier: ($) => "async",
 
-    access_modifier: ($) => choice("public", "private", "protected"),
+    access_modifier: ($) => choice("public", "protected"),
```
[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply libs/tree-sitter-wing/grammar.js:535]: Yes, it seems to me more strongly related to "compilation units" and I as such I'd like a word with @Chriscbr (our module guy) to better understand the design.

File: libs/wingc/src/type_check.rs (hunk 7 above)
Comment (yoav-steinberg) [reply libs/wingc/src/type_check.rs]: oops, forgot to delete this comment. This is a leftover comment from an earlier iteration. Resolved.

File: libs/wingc/src/type_check.rs (hunk 4 above)
Comment (yoav-steinberg) [reply libs/wingc/src/type_check.rs]: I changed to:
"Check if the class in which the property is defined is one of the classes we're currently nested in"
Hopefully that's clearer. I definitely struggled to correctly describe what I'm doing here (and I'm also not happy with the method name `self.ctx.current_class_nesting` 3 lines below).
What I'm doing here is I run up the stack of currently nested classes to see if the property access is from one of those classes. Here's an example to make it clearer:
```wing
class A {
  static foo() {} // This is a **private** static method.
  method() {
    class B {
        method() {
          A.foo() // I can access A's private stuff since I'm nested inside B which is nested inside A.
        }
    }
  }
}
```

File: libs/wingc/src/type_check.rs [hunk 9]
```diff
@@ -3799,43 +3943,7 @@ impl<'a> TypeChecker<'a> {
 							.collect_vec()[0]
 							.1;
 
-						let class_initializer = &class_type
```
[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply libs/wingc/src/type_check.rs:3802]: Thanks for looking into this. I see the issue. I mixed things up. I'll try to fix it.

[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply libs/wingc/src/type_check.rs:3802]: I pushed a fix, the gist of is is that if we type check the `super()` call statement during the type checking of the `init` itself then there's no need for env acrobatics. If we try to do this in advance then your original logic holds, with a slight required change, but it seems a bit complex. Can you review this commit: https://github.com/winglang/wing/pull/4133/commits/e36efd38f4d89ebed6e94762a0ea44f739220b3f

File: libs/wingc/src/type_check.rs (hunk 1 above)
Comment (yoav-steinberg) [reply libs/wingc/src/type_check.rs:3917]: resolving since we've continued this discussion in another conversation below

File: examples/tests/invalid/access_modifiers.w (hunk 6 above)
Comment (yoav-steinberg) [reply examples/tests/invalid/access_modifiers.w:4]: Good callout. Anything implementing an interface must be `public`. I'll need to added verification and tests...

File: libs/wingc/src/type_check.rs (hunk 4 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs]: Got it, yeah makes more sense now  - thanks!

File: examples/tests/invalid/access_modifiers.w (hunk 6 above)
Comment (yoav-steinberg) [reply examples/tests/invalid/access_modifiers.w:10]: https://github.com/winglang/wing/pull/4133/commits/d3d1c3c33f35fb43ed570f2881d5b8782821d5fa

File: libs/wingc/src/ast.rs [hunk 10]
```diff
@@ -513,6 +518,14 @@ pub struct ClassField {
 	pub reassignable: bool,
 	pub phase: Phase,
 	pub is_static: bool,
+	pub access_modifier: AccessModifier,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum AccessModifier {
```
[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply libs/wingc/src/ast.rs:525]: Added: https://github.com/winglang/wing/issues/4156

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 11]
```diff
@@ -10,6 +10,147 @@ exports[`access_hidden_namespace.w 1`] = `
...
+Test Files 1 failed (1)
+Duration <DURATION>"
+`;
+
+exports[`access_modifiers.w 1`] = `
+"error: Method \\"private_method\\" is not public but it's part of the \\"SomeInterface\\" interface
+   --> ../../../examples/tests/invalid/access_modifiers.w:40:3
+   |
+40 |   private_method() {}
+   |   ^^^^^^^^^^^^^^ Method \\"private_method\\" is not public but it's part of the \\"SomeInterface\\" interface
```
Comment (Chriscbr) [inline tools/hangar/__snapshots__/invalid.ts.snap:23]: optional - I'm not sure if this message makes it obvious why being part of the interface means it has to be public. Maybe we could reword this to `Interface members must be declared public` or `Method "foo" is private in "Foo" but not in "SomeInterface"`, or something like that.

```suggestion
40 |   private_method() {}
   |   ^^^^^^^^^^^^^^ Interface members must be declared public
```

File: libs/tree-sitter-wing/grammar.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply libs/tree-sitter-wing/grammar.js]: Added issue and referenced it diags: #4156

File: examples/tests/invalid/access_modifiers.w
[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply examples/tests/invalid/access_modifiers.w:1]: please review: https://github.com/winglang/wing/pull/4133/commits/5e7c8c71c81e62a74293eb479c4d8821659d33c8

[the remark this answers is not in the document]
Comment (Chriscbr) [reply examples/tests/invalid/access_modifiers.w:1]: LGTM

File: libs/wingc/src/type_check.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply libs/wingc/src/type_check.rs:3802]: 🤌

File: libs/wingsdk/src/.gen/versions.json
Comment (Chriscbr) [inline libs/wingsdk/src/.gen/versions.json:1]: I think this file is supposed to be gitignored now

**Review by Chriscbr — APPROVED** [review]
